### PR TITLE
Add intro to functional programming

### PR DESCRIPTION
**Pull request by cartermp**
First in a series that will be written over time.

### CODE REVIEW

File: docs/fsharp/introduction-to-functional-programming/index.md [hunk 1]
```diff
@@ -1,8 +1,221 @@
...
 ---
 
 # Introduction to Functional Programming in F# #
 
-> [!NOTE]
-This is still in-progress.
\ No newline at end of file
+Functional programming is a style of programming that emphasizes the use of functions and immutable data. Typed functional programming is when functional programming is combined with static types, such as with F#. In general, the following concepts are emphasized in functional programming:
+
+* Functions over objects
```
Comment (BillWagner) [inline docs/fsharp/introduction-to-functional-programming/index.md]: I like this list as a set of goals. You did a great job of explaining immutable values over variables. I didn't find any comparison of functions and objects, or declarative over imperative. I mentioned some suggestions to make the comparison of expressions and suggestions stronger.  

If your intended audience is developers familiar with OO or procedural languages, I'd make those comparisons more clear.

**Review by BillWagner — CHANGES_REQUESTED** [review]
Overall, this is well presented.  You've explained the concepts clearly, with great examples of F# syntax.

I had a few suggestions that would make the comparisons stronger if your intended audience is developers familiar with OO or procedural languages.

I'd also recommend running Acrolinx on these changes. I saw a few instances of passive voice and future tense. The tool is more efficient at finding those than I am.

File: docs/fsharp/introduction-to-functional-programming/index.md [hunk 2]
```diff
@@ -1,8 +1,221 @@
...
+Since any type in F# can have `ToString()` called on it, the type of `x` has been generalized to be anything, and the resultant type is a `string`.
+
+Expressions are not just the bodies of functions. You can have expressions that produce a value you use elsewhere. A common one is `if`:
+
+```fsharp
+// Checks if 'x' is odd by using the mod operator
+let isOdd x = x % 2 <> 0
+
+let addOneIfOdd input =
+    let result =
```
[the remark this answers is not in the document]
Comment (cartermp) [reply docs/fsharp/introduction-to-functional-programming/index.md]: I added more detail and an C# comparison.

File: docs/fsharp/introduction-to-functional-programming/index.md [hunk 3]
```diff
@@ -1,8 +1,221 @@
...
+            input + 1
+        else
+            input
+
+    result
+```
+
+The `if` expression produces a value called `result`. The key thing to remember about expressions is that they produce values in functional programming.
+
+### Pure functions
```
[the remark this answers is not in the document]
Comment (cartermp) [reply docs/fsharp/introduction-to-functional-programming/index.md]: Good idea. I've done so here and in the referential transparency section.

**Comment by mairaw** [comment]
@BillWagner this might be ready for a new review

**Review by BillWagner — APPROVED** [review]

**Comment by BillWagner** [comment]
I really like this @cartermp 
You can :shipit: when ready.

File: docs/fsharp/introduction-to-functional-programming/index.md [hunk 4]
```diff
@@ -1,8 +1,276 @@
...
+
+The signature looks like this:
+
+```fsharp
+val printString: string -> unit
+```
+
+The `unit` type indicates that there is no actual value being returned, and allows for you to specify something that conceptually does nothing.
+
+This is in sharp contrast to imperative programming, where the equivalent `if` construct is a statement, and producing values is often done with mutating variables. For example, in C#, the code might be written like this:
```
[the remark this answers is not in the document]
Comment (cartermp) [reply docs/fsharp/introduction-to-functional-programming/index.md]: Added.

**Review by svick — APPROVED** [review]

File: docs/fsharp/introduction-to-functional-programming/index.md [hunk 5]
```diff
@@ -1,8 +1,280 @@
...
+    if (IsOdd(input))
+    {
+        result = input + 1;
+    }
+
+    return result;
+}
+```
+
+It's worth noting that C# and other C-style languages do suppor the ternary `?` expression, which allows for expression-based conditional programming.
```
Comment (pkulikov) [inline docs/fsharp/introduction-to-functional-programming/index.md]: typo: "do suppor" -> "do support"
nit: `?` -> `?:` (not to mix with the null-conditional `?.` operator) or even explicitly: `cond ? x : y`

**Pull request merged**